### PR TITLE
fix(snapshots/thin): a thick volume becomes thin after snapshots are created

### PR DESCRIPTION
**Pull request by tiagolobocastro**
test: tests should clean themselves
    
    Signed-off-by: Tiago Castro <tiagolobocastro@gmail.com>

---

    fix(snapshot): update replica information
    
    After taking a snapshot the replica info changes, so update it as well.
    
    Signed-off-by: Tiago Castro <tiagolobocastro@gmail.com>

---

    feat(snapshot/usage): add snapshot usage for the replica usage
    
    In preparation of pending dataplane rpc changes.
    
    Signed-off-by: Tiago Castro <tiagolobocastro@gmail.com>

---

    fix(snapshots/thin): a thick volume becomes thin after snapshots are created
    
    After first snapshot is created we mark the volume as thin and this is sticky.
    
    Signed-off-by: Tiago Castro <tiagolobocastro@gmail.com>

### CODE REVIEW

File: control-plane/stor-port/src/types/v0/store/volume.rs [hunk 1]
```diff
@@ -428,9 +470,15 @@ impl SpecTransaction<VolumeOperation> for VolumeSpec {
 
     fn log_op(&self, operation: &VolumeOperation) -> (bool, bool) {
         match operation {
-            VolumeOperation::CreateSnapshot(_) | VolumeOperation::DestroySnapshot(_) => {
-                (false, false)
+            VolumeOperation::CreateSnapshot(_) => {
+                if !self.thin && !self.as_thin() {
```
Comment (dsharma-dc) [inline control-plane/stor-port/src/types/v0/store/volume.rs]: Is this check to be interpreted as `We are thick currently, and this is the first snapshot being created` ?

File: control-plane/stor-port/src/types/v0/store/volume.rs [hunk 2]
```diff
@@ -172,10 +172,44 @@ pub struct VolumeSpec {
...
+                snapshot_as_thin: as_thin,
+            },
+            runtime: Default::default(),
+        }
+    }
+    /// Insert snapshot in the list.
+    fn insert_snapshot(&mut self, snapshot: SnapshotId) {
+        self.runtime.snapshots.insert(snapshot);
+        // we become thin provisioned!
+        self.persisted.snapshot_as_thin = true;
```
Comment (dsharma-dc) [inline control-plane/stor-port/src/types/v0/store/volume.rs]: I think we'll be doing this for each snapshot creation but effectively only first snapshot is doing this? Shall we make it more intuitive like:
```
if !self.persisted.snapshot_as_thin
    self.persisted.snapshot_as_thin = true;
```

File: control-plane/stor-port/src/types/v0/store/volume.rs (hunk 1 above)
Comment (tiagolobocastro) [reply control-plane/stor-port/src/types/v0/store/volume.rs]: Yes correct, do you think self.thin && !self.has_snapshots() would be clearer?

File: control-plane/stor-port/src/types/v0/store/volume.rs (hunk 2 above)
Comment (tiagolobocastro) [reply control-plane/stor-port/src/types/v0/store/volume.rs]: hmm I don't see that as more intuitive, it's less to me before it's redundant.

**Review by Abhinandan-Purkait — APPROVED** [review]
Do we need plugin changes?

**Comment by tiagolobocastro** [comment]
bors merge

File: control-plane/stor-port/src/types/v0/store/volume.rs (hunk 1 above)
Comment (tiagolobocastro) [reply control-plane/stor-port/src/types/v0/store/volume.rs]: Used has_snapshots check.

**Comment by tiagolobocastro** [comment]
bors cancel

**Comment by tiagolobocastro** [comment]
bors merge

**Pull request merged**